test/writer_test.py

```diff
@@ -0,0 +1,239 @@
+    cfg.file_source_id = 200
+    writer = copc.FileWriter(file_path, cfg)
+
+    las_header = writer.GetLasHeader()
+    assert las_header.file_source_id == 200
+    assert las_header.point_format_id == 8
+    # TODO[Leo]: Update after making our own lazperf headers
+    # assert las_header.scale.x == 2
+    # assert las_header.offset.x == -0.02
+    # assert las_header.scale.y == 3
+    # assert las_header.offset.y == -0.03
+    # assert las_header.scale.z == 4
+    # assert las_header.offset.z == -40.8
+
+    writer.Close()
+
+    # COPC Span
+
+    cfg = copc.LasConfig(0)
+    writer = copc.FileWriter(file_path, cfg, 256)
+
+    # todo: use Reader to check all of these
+    assert writer.GetCopcHeader().span == 256
+
+    writer.Close()
+
+    reader = copc.FileReader(file_path)
+    assert reader.GetCopcHeader().span == 256
+
+    # WKT
+    cfg = copc.LasConfig(0)
+    writer = copc.FileWriter(file_path, cfg, 256, "TEST_WKT")
+
+    # todo: use Reader to check all of these
+    assert writer.GetWkt() == "TEST_WKT"
+
+    writer.Close()
+
+    reader = copc.FileReader(file_path)
+    assert reader.GetWkt() == "TEST_WKT"
+
+    # Copy
+    orig = copc.FileReader("../build/test/data/autzen-classified.copc.laz")
+
+    cfg = copc.LasConfig(orig.GetLasHeader(), orig.GetExtraByteVlr())
+    writer = copc.FileWriter(file_path, cfg)
+    writer.Close()
+
+    reader = copc.FileReader(file_path)
+    assert reader.GetLasHeader().file_source_id == orig.GetLasHeader().file_source_id
+    assert reader.GetLasHeader().global_encoding == orig.GetLasHeader().global_encoding
+    # assert reader.GetLasHeader().creation.day == orig.GetLasHeader().creation.day
+    # assert reader.GetLasHeader().creation.year == orig.GetLasHeader().creation.year
+    assert reader.GetLasHeader().file_source_id == orig.GetLasHeader().file_source_id
+    assert reader.GetLasHeader().point_format_id == orig.GetLasHeader().point_format_id
+    assert (
+        reader.GetLasHeader().point_record_length
+        == orig.GetLasHeader().point_record_length
+    )
+    assert reader.GetLasHeader().point_count == 0
+    # assert reader.GetLasHeader().scale == reader.GetLasHeader().scale
+    # assert reader.GetLasHeader().offset == reader.GetLasHeader().offset
+
+
+def test_writer_pages():
+    # Given a valid file path
+    file_path = "../build/test/data/writer_test.copc.laz"
+
+    # Root Page
+    writer = copc.FileWriter(file_path, copc.LasConfig(0))
+
+    assert not writer.FindNode(copc.VoxelKey.BaseKey()).IsValid()
+    assert not writer.FindNode(copc.VoxelKey.InvalidKey()).IsValid()
+    assert not writer.FindNode(copc.VoxelKey(5, 4, 3, 2)).IsValid()
+
+    writer.GetRootPage()
+    root_page = writer.GetRootPage()
+    assert root_page.IsValid()
+    assert root_page.IsPage()
+    assert root_page.loaded is True
+    with pytest.raises(RuntimeError):
+        writer.AddSubPage(root_page, copc.VoxelKey.InvalidKey())
+
+    writer.Close()
```

Comment:
scoped

Reply:
Not sure what you mean here?

Reply:
remove the Close and make the writer scoped